File: tests/io/foreign_shared.c
// A balanced tree of the depth asked, its leaves 1.. left to right, every
// odd one Empty; io_node seals the Node's fields as the program's own
// Node does, since the program shares the tree.
static Term shared_make_at(Env e, u32 depth, u32* next) {
  if (depth == 0) {
    *next += 1;
    return *next % 2 == 1 ? term_pak(CID_EMPTY, 0) : term_pak(CID_LEAF, *next);
  }
  Term l = shared_make_at(e, depth - 1, next);
  Term r = shared_make_at(e, depth - 1, next);
  return io_node(e, CID_NODE, l, r);
}

Term shared_make_run(Env e, Term* f, IoWork* w) {
  u32 next = 0;
  return shared_make_at(e, (u32)f[0], &next);
}

static void __attribute__((constructor)) foreign_shared_use(void) {
  io_eff(CID_SHARED_MAKE, shared_make_run, 0);
}

File: tests/io/foreign_shared.js
// A balanced tree of the depth asked, its leaves 1.. left to right, every
// odd one Empty, as the C lays it.
function shared_make(depth) {
  let next = 0;
  const at = (d) => {
    if (d === 0) {
      next += 1;
      return next % 2 === 1 ? { $: "Empty" } : { $: "Leaf", v: next };
    }
    const l = at(d - 1);
    return { $: "Node", l, r: at(d - 1) };
  };
  return at(depth);
}
